### ISSUE TITLE
Add support for federation

### ISSUE THREAD

**Issue by phillip-kruger** [smallrye/smallrye-graphql#521]
Might be a lot of work, but let's consider support for federation

https://www.apollographql.com/docs/federation/managed-federation/overview/

**Comment by chet20r** [comment]
Thanks Phillip, I think Federation support would be really helpful given organizations adopting microservices pattern. Especially with large organizations with varied technology stacks who want to have a single unified data graph combining multiple microservices on different technology platforms.

There is already a support for federation on the JVM via apollo [federation-jvm](https://github.com/apollographql/federation-jvm) package. Probably worth looking if that can be leveraged here?

**Comment by t1** [comment]
This is about services taking part in a federation, right? Writing a federation gateway is a completely different and much bigger topic. I've read several blogs and watched several talks trying to understand esp. the difference between Schema Stitching and Federation. For me, [this](https://www.apollographql.com/blog/the-evolution-of-graphql-at-scale/) blog post explained it best.

My ultra-short summary: Schema Stitching works with custom code in the Gateway that hooks the endpoints together. While this approach starts very simple, complexity grows exponentially, as there is a single point that has to know about _all_ connections. Federation OTOH distributes this responsibility to the separate services, e.g. a user service only knows about the `User` type, while a review service _extends_ the `User` type with a `reviews: [Review!]!` field while only knowing about the id of the user. This and a few more things is done with quite flexible directives; and the gateway then does all the Federation magic on a generic level.

**Comment by t1** [comment]
Maybe it would be sufficient to create a separate library like `smallrye-graphql-federation` that contains the required types and some [custom directives](https://github.com/eclipse/microprofile-graphql/issues/344)?

**Comment by t1** [comment]
The correct link to start from is https://www.apollographql.com/docs/federation/federation-spec/

The managed-federation link in the description is about the advanced federation gateway in the Apollo Studio product that manages the services with a registry instead of having to list them manually before startup of the gateway.

**Comment by t1** [comment]
I found that if we really want to support federation with an extension (i.e. a library you can include, not something built directly into the server implementation), we not only need custom directives (#628 for `@external`, `@key`, `@provides`, `@requires`, and `@extends`), we also need:

* Unions (#400 for `_Entity`); this even has to be dynamic, as the union is the list of all types with a `@key` directive.
* Custom scalars ([MP#94](https://github.com/eclipse/microprofile-graphql/issues/94) for `_Any` and `_FieldSet`); `_Any` is a complex scalar (which is an oxymoron in my eyes).

**Comment by t1** [comment]
I've started some experiments and pushed them [here](https://github.com/t1/graphql-federation-java-experiment). There are many prio A TODOs left, but it works as a PoC!

I'd love to hear your feedback.

**Comment by phillip-kruger** [comment]
Thanks @t1 ! I'll have a look a.s.a.p

**Comment by t1** [comment]
As mentioned in the meeting yesterday, there is one major issue open: a type used as a `@FederatedSource` argument is not automatically added to the schema, so as a workaround I have to define something like `@Query public Film dummyFilm() { return null; }`.

**Comment by t1** [comment]
It could be possible that #312 could help to fix this last priority A todo.

**Comment by phillip-kruger** [comment]
Can we maybe discuss here (again) why can we not use `@Source` ?

**Comment by t1** [comment]
> why can we not use @Source ?

You mean instead of `@FederatedSource`? I actually haven't thought about that before. It could work and fix the problem with registering the type. But we'd have to replace the local resolver logic. I'm not sure how to do that.

**Comment by phillip-kruger** [comment]
Yea, so the class needs an `@External` right ? So we should be able to figure it out, but it might need some changes in SR impl module

**Comment by mistreckless** [comment]
What is current status for the federation support? Can I use it with quarkus in expiremental/alfa status on my own risk? There are still no federation quarkus extension [#15386]( https://github.com/quarkusio/quarkus/issues/15386 ).

**Comment by t1** [comment]
It currently works as an alpha in JEE containers, but not in Quarkus. We're still working on the Quarkus Extension, but there was not much progress lately. Do you want to support?

**Comment by shmulik-klein** [comment]
@t1 I would really like to help with the Quarkus extension, where can I start from?

**Comment by t1** [comment]
Maybe @phillip-kruger can give you a hint?

**Comment by phillip-kruger** [comment]
I'll be honest - I am not sure. I am not sure if this would be a new extension, or build into the current one. I think the current one. So maybe start by playing around with that ?

**Comment by shmulik-klein** [comment]
Thanks @phillip-kruger.

@t1 @phillip-kruger Just to make sure that I understand it correctly - `quarkus-smallrye-graphql` Quarkus extension, generates the graphql schema at build time (deployment), so it can't work with `smallrye-graphql-federation-runtime` as the later observes to the schema creation and adds to it the entities required to make it a federated schema? - meaning, you can only use it entirely as a Quarkus extension or entirely as a regular dependency, but can't make a mix of those?

**Comment by shmulik-klein** [comment]
Another small question (+@t1), is it possible to use `smallrye-graphql` + `smallrye-graphql-federation-runtime` dependencies on quarkus to get graphql federation support? I tried to do so with `undertow` + `smallrye-graphql-servlet`, but failed. I tried to do so to better understand the bootstrapping logic.

**Comment by phillip-kruger** [comment]
@shmulik-klein @t1 - I think we said that we should merge the federation stuff (in SmallRye) into the smallrye-implementation. That means we need to move the annotation discovery to the common/modelbuilder module. That way, quarkus should just work

**Comment by shmulik-klein** [comment]
My intuition was to combine the extension, but this sounds better. I'll look into that.

**Comment by t1** [comment]
Please note that the main thing missing is tests. They'd have to live in the SmallRye TCK.

We also talked about using the `@Source` annotation instead of the `@FederatedSource`, which might let us get rid of the requirement to declare federated types as a query result in order to use them as a federated source.

**Comment by t1** [comment]
@shmulik-klein: It would be cool if you could share your progress as early and as often as possible in a PR, as I need this, too.

**Comment by shmulik-klein** [comment]
@t1 I had to overcome the following issue #963 to be able to run the maven plugin and to better understand the `schema-builder` flow. 
I understand that the schema builder generates a `io.smallrye.graphql.schema.model.Schema`, while the `Federation`'s code acts on `graphql.schema.GraphQLSchema.Builder`. In order to integrate it into `common/schema-builder` should I refactor `Federation` to act on `Schema` or is there a better way around?

Also, `Federation` relies on CDI, while for example, the maven plugin doesn't provide such, do we also want to remove the CDI dependency from `Federation`?

I'll try to get a PR asap, but I'm still trying to understand the flow and the right way to integrate it.

**Comment by t1** [comment]
I haven't thought about this earlier. The existing federation module was planned as a "plugin", so it does everything at runtime. When integrating it into the core module, there are some static parts (_service and _entities) and some dynamic parts that depend on the other types used (_Entity).

I think we need a config option to enable it.

Creating a PR is helpful to have code to discuss about, even when it's still completely in progress. You can mark the PR as explicitly "draft".

**Comment by emmanuelsilva** [comment]
Hi @t1, the implementation looks ok on the smallrye-graphql side.

I installed this branch locally, and I tried to test it using Quarkus by creating an extension, but I couldn't make it work due to issue #962. The code is in this [github repository](https://github.com/emmanuelsilva/quarkus-graphql-federation-api)

So, as far as I understood, we need to wait for fixing that quarkus issue to test the quarkus integration.

**Comment by t1** [comment]
I had blindly assumed that the existing quarkus extension would work with this new approach to federation out of the box. What did you have to add?

**Comment by emmanuelsilva** [comment]
@t1, using the current quarkus extension was my first attempt, like that:

```xml
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-smallrye-graphql</artifactId>
</dependency>
<dependency>
    <groupId>io.smallrye</groupId>
    <artifactId>smallrye-graphql</artifactId>
    <version>1.4.0-SNAPSHOT</version>
</dependency>
```

Is it correct? I've updated the [repository](https://github.com/emmanuelsilva/quarkus-graphql-federation-api)

I also came across the exception described in #962 using this configuration.

**Comment by emmanuelsilva** [comment]
I changed [my example](https://github.com/emmanuelsilva/quarkus-graphql-federation-api) to use the already existent quarkus extension:

```xml
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-smallrye-graphql</artifactId>
</dependency>
<dependency>
    <groupId>io.smallrye</groupId>
    <artifactId>smallrye-graphql</artifactId>
    <version>1.4.0-SNAPSHOT</version>
</dependency>
```

Here are my findings:

1 - I needed to enable the federation by changing directly the method `io.smallrye.graphql.spi.config.Config.isFederationEnabled()` to return `true` because I set `smallrye.graphql.federation.enabled=true` in the application.properties, but it didn't enable the federation.

 Do you know how to configure it using the application.properties?


2 - After enabling the federation, I could see the `sdl` in the graphql schema:

```graphql
type _Service {
  sdl: String!
}
```


3 - I couldn't see the federation directives such as `@key` `@extends` in the graphql schema even the property `quarkus.smallrye-graphql.schema-include-directives=true` is enabled.

**Comment by t1** [comment]
1 - The Quarkus Extension does some magic with the config, that I'm not familiar with. Maybe @phillip-kruger can shed some light.

3 - Is that in the "normal" schema or in the schema returned from `_service { sdl }`?

**Comment by phillip-kruger** [comment]
I'll have a look a.s.a.p

**Comment by R4V3NN0** [comment]
We are trying to integrate our service into an apollo gateway. We have tried to use some of the examples from @t1  and @emmanuelsilva but we could not make it work.
Is federation in general already possible and we are just having a bad time to make it work or is there something that must be coded before?

**Comment by phillip-kruger** [comment]
It's not 100% supported yet and on our todo list to make it work properly. Maybe @t1 can help to get yours to work

**Comment by t1** [comment]
I'm currently very busy (in my free time) with https://github.com/t1/wunderbar, but federation will come up in my current project, so I will resume working on it soon. Your example may be a good starting point.

**Comment by robp94** [comment]
@R4V3NN0 we have it working, but it is ugly. Have a look at this https://github.com/smallrye/smallrye-graphql/pull/1135
Apart from an additional jandex index, we although change the schema of our services in the gateway to add the key fields and other annotations.

**Comment by eformat** [comment]
Hi .. not sure how this is going, but can't wait till quarkus supports federation!

i had some success compiling an example against @t1 's branch - https://github.com/t1/smallrye-graphql/tree/521-federation-into-core which seems like a WIP - using the apollo gateway and quarkus graphql backend's. One thing that i was stuck on for quite some time turned out to be something pretty trivial. I continually got errors from the gateway like this:

```
Error: Couldn't load service definitions Expected JSON response body, but received: {"data":{"_service":{"sdl":"schema {\n  query: Query\n}\n\ntype Movie {\n  movieId: String!\n  title: String!\n}\n\n\"Query root\"\ntype Query {\n  movie(movieId: String!): Movie\n}\n"}}}
```

turns out quarkus graphql returns a mimeType of <strong>application/graphql+json; charset=UTF-8</strong>
and apollo gateway is only checking for <strong>application/json<strong> e.g. i modified this

https://github.com/apollographql/federation/blob/main/gateway-js/src/datasources/RemoteGraphQLDataSource.ts#L282

to be

```
    parseBody(fetchResponse, _fetchRequest, _context) {
        const contentType = fetchResponse.headers.get('Content-Type');
        if (contentType && (contentType.startsWith('application/json') || contentType.startsWith("application/graphql+json"))) {
            return fetchResponse.json();
        }
        else {
            return fetchResponse.text();
        }
    }
```

hope that helps someone else watching this ! thanks.

**Comment by phillip-kruger** [comment]
Nice ! Thanks @eformat .

**Comment by t1** [comment]
Super cool. Maybe we should make the content type that Smallyre returns a config option?

**Comment by phillip-kruger** [comment]
Agree, at least the default one as fallback, first looking at matching an accept header ?

**Comment by t1** [comment]
@eformat : do you happen to know what `Accept` header Apollo sends? We could try to match that.

**Comment by eformat** [comment]
wiresharking the connection from the apollo gateway to the /graphql backend seems to suggest it sets

```
Accept: */*
```

but then does not handle it in the code :)

![Screenshot from 2022-04-22 17-19-33](https://user-images.githubusercontent.com/712608/164626171-0b15ca43-b5d4-410e-8ec7-ac9bcdfb97dd.png)

**Comment by phillip-kruger** [comment]
We follow the [GraphQL-over-HTTP Spec](https://github.com/graphql/graphql-over-http), that is why we respond with `application/graphql+json` (see https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#content-types) 

If the request contains an accept header that indicate `application/json` that is what we will return with. We have no rule for `*/*` hence it fallback to the default.

I think that is the correct default, but what we can do is to allow user to configure the response content type header for `*/*`. 
I also think it's a bug on Apollo side to request `*/*`, so another option is to open an issue/PR there ?

@eformat - please open an new issue that request the config for `*/*` if this will work for you.

**Comment by eformat** [comment]
yeah .. i'm going to go down the Apollo approach and try a PR over there, that seems the logical place to fix it

**Comment by eformat** [comment]
lets see how this goes

https://github.com/apollographql/federation/pull/1767

**Comment by R4V3NN0** [comment]
At the moment I am trying to integrate federation with a combination from `federation-jvm` and `graphql-java`.
I tried to use the leaky abstraction, as it is described in https://quarkus.io/guides/smallrye-graphql#graphql-java and then I wanted to transform the schema with `federation-jvm` as it is described in their Github project https://github.com/apollographql/federation-jvm as a result I have implemented a minimal example https://github.com/R4V3NN0/quarkus-smallrye-graphql-federation-example/blob/main/src/main/java/configuration/FederationEnabler.java but at the moment it is not working. Something with the leaky abstraction seems not to work alright... 

Has anyone ever tried an approach like this or is this approach totally nonsense?

**Comment by phillip-kruger** [comment]
Not sure. I will look at federation soon, and see if we can not make it part of the lib without any effort from the user side.

**Comment by bizzi-vinhvo** [comment]
Linked Issue: https://github.com/quarkusio/quarkus/issues/26722

**Comment by ntkoopman** [comment]
What's the status of this? Is there anything I can help with?

**Comment by phillip-kruger** [comment]
Hi @ntkoopman ! Thanks for offering to help. As far as I know basic support is available in Quarkus. The outstaning items is getting the reactive part to work correctly (or at all). So interpret Multi/Uni and execute those on the eventloop. @t1 @jmartisk can you comment ?

**Comment by MightyPixel** [comment]
Hi everyone!
I've saw [this](https://www.apollographql.com/docs/federation/building-supergraphs/supported-subgraphs/) page in the apollo website. Does it make sense to add the a table for `smallrye-graphql` to keep track what is supported?

**Comment by t1** [comment]
@phillip-kruger & @ntkoopman: Sorry for not responding before. We currently don't use it in my project, so I can't invest a lot of time on this at the moment. Your summary of the current status looks quite like what I remember it to be. I actually tried to fix the reactive stuff, but gave up. I suppose my brain is just not made to work with reactive logic. I'm soooo glad that Project Loom is coming to save me 😁

@MightyPixel: That would be great. We would need to run the specified tests against our implementation. That would be an essential step forward. Maybe you can get them running (and failing)?

**Comment by robp94** [comment]
Reactive should work at least Unis.
